File: lib/crane/models/service.rb

require "digest/md5"
require "yaml"
require "fileutils"
require "uri"
require "pathname"

module Crane
    
    module Models
        
        class Service
            
            def initialize(infrastructure, serviceName, config)
                
                folderName = File.dirname infrastructure.file
                
                if config.has_key? "manifest"
                    
                    @file = config["manifest"]
                    
                    if !(@file =~ URI::regexp)
                        @file = "#{folderName}/#{config['manifest']}"
                    end
                    
                else
                    
                    @file = "#{folderName}/#{serviceName}.yml"
                    
                end
                
                raw = YAML.load_file @file
                
                Crane.hashToObject self, raw
                
            end
            
            def getRelativePath(checkoutDirectory, serviceManifest, source)
                
                relativeManifestPath = serviceManifest.relative_path_from checkoutDirectory
                relativeFolderPath = Pathname.new File.dirname(relativeManifestPath.to_s)
                relativeSourcePath = Pathname.new "#{relativeFolderPath.to_s}/#{source}"
                
                relativeSourcePath
                
            end
            
            def addPuppetFacts(file, checkoutPath, serviceManifest)
                
                provision.facts.each do |fact|
                    
                    sourcePath = getRelativePath checkoutPath, serviceManifest, fact
                    file.puts "ADD #{sourcePath.to_s} /etc/facter/facts.d/"
                    
                end
                
            end
            
            def addPuppetModules(file, checkoutPath, serviceManifest)
                
                index = 0
                provision.modulePaths.each do |moduleFolder|
                    
                    sourcePath = getRelativePath checkoutPath, serviceManifest, moduleFolder
                    file.puts "ADD #{sourcePath.to_s} /tmp/puppet/_modules-#{index}/"
                    index = index + 1
                    
                end
                
            end
            
            def addPuppetManifest(file, checkoutPath, serviceManifest)
                
                sourcePath = getRelativePath checkoutPath, serviceManifest, provision.manifest
                file.puts "ADD #{sourcePath} /tmp/puppet/_manifest/manifest.pp"
                
            end
            
            def applyPuppetProvisioning(file, checkoutPath, serviceManifest)
                
                addPuppetFacts file, checkoutPath, serviceManifest
                addPuppetModules file, checkoutPath, serviceManifest
                addPuppetManifest file, checkoutPath, serviceManifest
                
                # http://docs.puppetlabs.com/references/stable/configuration.html#modulepath
                file.puts "RUN puppet apply --modulepath=#{modules.join(':')} /tmp/puppet/_manifest/manifest.pp"
                
            end
            
            def applyShellProvisioning(file, checkoutPath, serviceManifest)
                
                provision.directories.each do |source, target|
                    
                    sourcePath = getRelativePath checkoutPath, serviceManifest, source
                    
                    file.puts "ADD #{sourcePath.to_s} #{target}"
                    
                end
                
                if !provision.path.empty?
                    
                    file.puts "RUN PATH=#{provision.path.join(':')}:$PATH"
                    
                end
                
                provision.commands.each do |command|
                    
                    file.puts "RUN #{command}"
                    
                end
                
            end
            
            def copyCheckoutFolder(workingCopy, folder)
                
                Dir.foreach workingCopy.checkoutFolder do |file|
                    
                    if file != "." && file != ".."
                        
                        FileUtils::cp_r "#{workingCopy.checkoutFolder}/#{file}", folder
                        
                    end
                    
                end
                
            end
            
            def calculateFolder(workingCopy, environment)
                
                folder = Digest::MD5.hexdigest "#{workingCopy.project.name}-#{workingCopy.name}-#{name}-#{environment}"
                folder = "#{workingCopy.imageFolder}/#{folder}"
                
                unless File.directory? folder
                    
                    FileUtils.mkdir_p folder
                        
                end
                
                folder
                
            end
            
            def applyEnvironmentVariables(file, variables)
                
                variables.each do |key, value|
                    
                    file.puts "ENV #{key} #{value}"
                    
                end
                
            end
            
            def applyProvisioning(file, checkoutPath, filePath)
                
                if provision
                    
                    if provision.provider == "puppet"
                        
                        applyPuppetProvisioning file, checkoutPath, filePath
                        
                    elsif provision.provider == "shell"
                        
                        applyShellProvisioning file, checkoutPath, filePath
                        
                    end
                    
                end
                
            end
            
            def generateDockerfile(workingCopy, environment, variables)
                
                folder = calculateFolder workingCopy, environment
                dockerfile = "#{folder}/Dockerfile"
                
                copyCheckoutFolder workingCopy, folder
                
                checkoutPath = Pathname.new workingCopy.checkoutFolder
                filePath = Pathname.new @file
                
                File.open dockerfile, "w:UTF-8" do |file|
                    
                    file.puts "# #{name} (#{version})"
                    file.puts "FROM #{base}"
                    
                    applyEnvironmentVariables file, variables
                    applyProvisioning file, checkoutPath, filePath
                    
                    file.puts "EXPOSE #{ports.join(' ')}"
                    
                end
                
                dockerfile
                
            end
            
        end
        
    end
    
end

File: lib/models/service.js

var _ = require('underscore');
var Backbone = require('backbone');
var YAML = require('js-yaml');

module.exports.Service = Backbone.Model.extend({
    buildEnvironmentVariables: function(dockerfile, variables) {
        for(var key in variables) {
            dockerfile.push('ENV ' + key + ' ' + variables[key]);
        }
    },
    buildPuppetFacts: function(dockerfile) {
        var provision = this.get('provision');

        if(!!provision.facts) {
            for(var i = 0; i < provision.facts.length; i++) {
                dockerfile.push('ADD ' + provision.facts[i] + ' /etc/facter/facts.d/');
            }
        }
    },
    buildPuppetModules: function(dockerfile, modules) {
        var index = 0;
        var provision = this.get('provision');

        if(!!provision.modulePaths) {
            for(var i = 0; i < provision.modulePaths.length; i++) {
                var path = '/tmp/puppet/_modules-' + index++ + '/';
                dockerfile.push('ADD ' + provision.modulePaths[i] + ' ' + path);
            }
        }
    },
    buildPuppetManifest: function(dockerfile) {
        var provision = this.get('provision');
        dockerfile.push('ADD ' + provision.manifest + ' /tmp/puppet/_manifest/manifest.pp');
    },
    buildPuppetProvisioning: function(dockerfile) {
        var modules = [];

        this.buildPuppetFacts(dockerfile);
        this.buildPuppetModules(dockerfile, modules);
        this.buildPuppetManifest(dockerfile);

        // http://docs.puppetlabs.com/references/stable/configuration.html#modulepath
        dockerfile.push('RUN puppet apply --modulepath=' + modules.join(':') + ' /tmp/puppet/_manifest/manifest.pp');
    },
    buildShellProvisioning: function(dockerfile) {
        var provision = this.get('provision');

        if(!!provision.directories) {
            for(var i = 0; i < provision.directories.length; i++) {
                var dir = provision.directories[i];
                dockerfile.push('ADD ' + dir + ' ' + dir);
            }
        }

        if(!!slf.provision.path) {
            dockerfile.push('RUN PATH=' + provision.path.join(':') + ':$PATH');
        }

        if(!!provision.commands && provision.commands.length > 0) {
            for(var i = 0; i < provision.commands.length; i++) {
                dockerfile.push('RUN ' + provision.commands[i]);
            }
        }
    },
    buildProvisioning: function(dockerfile) {
        var provision = this.get('provision');

        if(!!provision) {
            if(provision.provider === 'puppet') {
                str = this.buildPuppetProvisioning(dockerfile);
            } else if(provision.provider === 'shell') {
                str = this.buildShellProvisioning(dockerfile);
            }
        }
    },
    buildDockerfile: function(options) {

        var dockerfile = [];

        dockerfile.push('# ' + this.get('name') + ' (' + this.get('version') + ')');
        dockerfile.push('FROM ' + this.get('base'));

        if(!!this.get('ports')) {
            dockerfile.push('EXPOSE ' + this.get('ports').join(' '));
        }

        this.buildEnvironmentVariables(dockerfile, options.variables);
        this.buildProvisioning(dockerfile);

        this.set('environments', this.get('environments') || {});
        this.get('environments')[options.environment] = {
            'dockerfile': dockerfile.join('\n')
        };

        options.callback();
    }
});
